src/io/parsers/GROParser.js

```diff
@@ -0,0 +1,230 @@
+    /* Adding residue and atom to complex structure */
+    const type = Element.getByName(this._atomName[0]); /* MAGIC 0. REASONS: This name is something like "CA", where
+     C - is an element an A is something else. But what about Calcium? */
+    if (type.fullName === 'Unknown') {
+      this._complex.error = {
+        message: `${this._atomName[0]} hasn't been recognised as an atom name.`,
+      };
+      return;
+    }
+    const role = Element.Role[this._atomName];
+    /* Firstly, create a dummy chain */
+    let chain = this._chain;
+    if (!chain) {
+      this._chain = chain = this._complex.addChain('A');
+    }
+    /* Secondly, add residue to that chain */
+    let residue = this._residue;
+    if (!residue || residue.getSequence() !== this._residueNumber) {
+      this._residue = residue = chain.addResidue(this._residueName, this._residueNumber, 'A');
+    }
+    /* Lastly, add atom to that residue */
+    this._atomPosition = new THREE.Vector3(positionX, positionY, positionZ);
+    /* Adding default constants to correct atom addition process */
+    const het = true;
+    const altLoc = ' ';
+    const occupancy = 1;
+    const tempFactor = 1;
+    const charge = 0;
+    residue.addAtom(this._atomName, type, this._atomPosition, role, het, this._atomNumber, altLoc, occupancy, tempFactor, charge);
+  }
+
+  /**
+   * Needed procedure for molecules finalization. In '.gro' file format there is only 1 chain and 1 molecule.
+   */
+  _finalizeMolecules() {
+    // get chain from complex
+    const chainDict = {};
+    const chain = this._complex._chains[0];
+    const chainName = chain._name;
+    chainDict[chainName] = chain;
+
+    // aggregate residues from chain
+    const m = this._molecules[0];
+    let residues = [];
+    residues = residues.concat(chain._residues.slice());
+
+    const molecule = new Molecule(this._complex, m._name, 1);
+    molecule._residues = residues;
+    this._complex._molecules[0] = molecule;
+  }
+
+  /**
+   * Some finalizing procedures.
+   * @returns {Complex} Complex structure for visualizing.
+   */
+  _finalize() {
+    this._molecule = { _index: '', _chains: [] };
+    this._molecule._index = 1;
+    this._molecules.push(this._molecule);
+    this._finalizeMolecules();
+    this._complex.finalize({
+      needAutoBonding: true,
+      detectAromaticLoops: this.settings.now.aromatic,
+      enableEditing: this.settings.now.editing,
+      serialAtomMap: this._serialAtomMap,
+    });
+  }
+
+  /**
+   * Main parsing procedure.
+   * @returns {Complex} Complex structure for visualizing.
+   */
+  parseSync() {
+    /* Create "Complex" variable */
+    const result = this._complex = new Complex();
+    /* Parse input file line-by-line */
+    const reader = new GROReader(this._data);
+    let counter = 0; /* Simple counter regarding to format of .gro file */
+    while (!reader.end()) {
+      /* First two lines - technical information, other lines - Atoms */
+      if (counter === 0) {
```

Comment:
Shouldn't we just make the first two calls and the last one outside the loop ("unroll it")? Probably, use the finite `for` loop for each atom and just check that the file is not over (just in case)?